core/src/HIP/Kokkos_HIP_Atomic.hpp

```diff
@@ -120,138 +154,86 @@ __inline__ __device__ void atomic_assign(
   atomic_exchange(dest, val);
 }
 
-inline __device__ int atomic_exchange(int *dest, const int &val) {
-  return atomicExch(dest, val);
-}
-
-inline __device__ unsigned int atomic_exchange(unsigned int *dest,
-                                               const unsigned int &val) {
-  return atomicExch(dest, val);
-}
-
-inline __device__ unsigned long long int atomic_exchange(
-    unsigned long long int *dest, const unsigned long long int &val) {
-  return atomicExch(dest, val);
-}
-
-inline __device__ float atomic_exchange(float *dest, const float &val) {
-  return atomicExch(dest, val);
-}
-
-template <class T>
-inline __device__ T atomic_exchange(T * /*dest*/, const T &val) {
-  // FIXME
-  Kokkos::abort("atomic_exchange not implemented!\n");
-  return val;
-}
+// atomic_compare_exchange -----------------------------------------------------
 
-inline __device__ int atomic_compare_exchange(int *dest, int compare,
+inline __device__ int atomic_compare_exchange(volatile int *dest, int compare,
                                               const int &val) {
-  return atomicCAS(dest, compare, val);
+  return atomicCAS(const_cast<int *>(dest), compare, val);
 }
 
 inline __device__ unsigned int atomic_compare_exchange(
-    unsigned int *dest, unsigned int compare, const unsigned int &val) {
-  return atomicCAS(dest, compare, val);
+    volatile unsigned int *dest, unsigned int compare,
+    const unsigned int &val) {
+  return atomicCAS(const_cast<unsigned int *>(dest), compare, val);
 }
 
 inline __device__ unsigned long long int atomic_compare_exchange(
-    unsigned long long int *dest, unsigned long long int compare,
+    volatile unsigned long long int *dest, unsigned long long int compare,
     const unsigned long long int &val) {
-  return atomicCAS(dest, compare, val);
+  return atomicCAS(const_cast<unsigned long long int *>(dest), compare, val);
 }
 
-template <typename T>
-inline __device__ T atomic_compare_exchange(T * /*dest*/, T /*compare*/,
-                                            const T &val) {
-  // FIXME
-  Kokkos::abort("atomic_compare_exchange not implemented!\n");
-  return val;
+template <class T>
+__inline__ __device__ T atomic_compare_exchange(
+    volatile T *dest, T compare,
+    typename std::enable_if<sizeof(T) == sizeof(int), const T &>::type val) {
+  union U {
+    int i;
+    T f;
+    __inline__ __device__ U() {}
+  } idest, icompare, ival;
+  idest.f    = *dest;
+  icompare.f = compare;
+  ival.f     = val;
+  idest.i    = atomicCAS((int *)(dest), icompare.i, ival.i);
```

Comment:
You should use a `reinterpret_cast` instead of the C cast. 
I don't get why you need an union here. You have already checked that `sizeof(T) == sizeof(int)` so what's the point. Also according to [this](https://en.cppreference.com/w/cpp/language/union) (It's undefined behavior to read from the member of the union that wasn't most recently written), you are relying on undefined behavior (`icompare.i` and `ival.i` are UB)

Reply:
I am really just following the `CUDA` implementation here.

Reply:
reinterpret_cast